File: vscode-extension/src/kotlinIcStorageDetector.ts
/**
 * Scans Gradle output for the Kotlin incremental-compiler "Storage already registered" failure
 * (upstream KT-59321 / KT-55435). The in-process Kotlin compile service retains a
 * `FilePageCache` registration across builds; a subsequent build's attempt to re-register the
 * same `caches-jvm/.../source-to-classes.tab` path throws.
 *
 * Symptom path that brought us here (issue #1493): editing a file under VS Code triggers
 * `gradleService.compileOnly` → daemon kicks `:<module>:composePreviewCompile`. Gradle reports
 * BUILD SUCCESSFUL because Kotlin falls back to non-incremental, but the on-disk IC cache is
 * left inconsistent, and the next save's `compileKotlin UP-TO-DATE` doesn't pick up the edit.
 * From the user's side, "save isn't doing anything."
 *
 * Feed each `onOutput` chunk through {@link consume}; after the task completes call
 * {@link end} and {@link getDetectedCachesJvmDirs}. A non-empty result is the signal to wipe
 * those directories, stop the Gradle daemon (which restarts the in-process Kotlin compile
 * service with a clean cache registry on next connect), and retry the build with
 * `--rerun-tasks`. The Kotlin-side equivalent in `:gradle-preview-driver`
 * (`KotlinIcStorageDetector.kt`) covers the CLI driver path; this module covers the
 * VS Code → vscjava.vscode-gradle daemon-compile path that the issue title names.
 */

import * as path from "path";

const STORAGE_RE = /Storage for \[([^\]]+)\] is already registered/;
const MAX_BUFFER = 16 * 1024;

/**
 * Walks the parents of [filePath] until a directory named `caches-jvm` is found. Returns null
 * if no such ancestor exists — defensive against future Kotlin error-message refactors that
 * change the cache layout. Exported for test access.
 */
export function findCachesJvmAncestor(filePath: string): string | null {
    let p = path.normalize(filePath);
    while (true) {
        const base = path.basename(p);
        if (base === "caches-jvm") return p;
        const parent = path.dirname(p);
        if (parent === p) return null;
        p = parent;
    }
}

export class KotlinIcStorageDetector {
    private buffer = "";
    private readonly detected = new Set<string>();

    /**
     * Accept a chunk of decoded stdout/stderr. Safe to call with partial lines — bytes are
     * buffered until a newline arrives. Unlike the single-finding detectors, this keeps
     * scanning after a hit so multi-module builds report every affected directory in one pass.
     */
    consume(chunk: string): void {
        this.buffer += chunk;
        let nl = this.buffer.indexOf("\n");
        while (nl !== -1) {
            const line = this.buffer.slice(0, nl);
            this.buffer = this.buffer.slice(nl + 1);
            this.scanLine(line);
            nl = this.buffer.indexOf("\n");
        }
        // Bound the buffer for pathological no-newline producers — 16 KiB dwarfs any single
        // Kotlin diagnostic and matches JdkImageErrorDetector's policy for consistency.
        if (this.buffer.length > MAX_BUFFER) {
            this.scanLine(this.buffer);
            this.buffer = "";
        }
    }

    /** Flush the residual buffer. Call once after the stream ends. */
    end(): void {
        if (this.buffer.length > 0) {
            this.scanLine(this.buffer);
            this.buffer = "";
        }
    }

    /** Distinct caches-jvm directories the recovery pass needs to wipe. */
    getDetectedCachesJvmDirs(): string[] {
        return Array.from(this.detected);
    }

    private scanLine(line: string): void {
        const m = STORAGE_RE.exec(line);
        if (!m) return;
        const cachesJvm = findCachesJvmAncestor(m[1]);
        if (cachesJvm) this.detected.add(cachesJvm);
    }
}

File: gradle-preview-driver/src/main/kotlin/ee/schimke/composeai/cli/KotlinIcStorageDetector.kt
package ee.schimke.composeai.cli

import java.io.File
import java.io.OutputStream
import java.util.Collections

/**
 * Tees an [OutputStream] (Gradle stdout/stderr) and scans the byte stream for Kotlin's incremental
 * compiler "Storage for [...] is already registered" failure. The marker is emitted by
 * `FilePageCache.registerPagedFileStorage` when an in-process Kotlin compiler daemon retains a
 * cache-file registration across builds (upstream KT-59321 / KT-55435).
 *
 * The Gradle build typically still reports BUILD SUCCESSFUL because Kotlin falls back to
 * non-incremental, but the on-disk `caches-jvm` directory can be left inconsistent and subsequent
 * `compileKotlin UP-TO-DATE` runs will not reflect the user's edit. [GradleConnection] uses the
 * detected directories to drive a stop-daemon + wipe-cache + retry recovery pass — see issue #1493.
 *
 * Detection is path-based: the marker carries an absolute path ending in
 * `caches-jvm/jvm/kotlin/source-to-classes.tab` (or a sibling `.tab` file). We walk parents until
 * we hit a directory named `caches-jvm` and record that as the wipe target.
 */
class KotlinIcStorageDetector(private val delegate: OutputStream) : OutputStream() {
  private val lineBuffer = StringBuilder()
  private val detected = Collections.synchronizedSet(linkedSetOf<File>())

  fun detectedCachesJvmDirs(): Set<File> = synchronized(detected) { detected.toSet() }

  override fun write(b: Int) {
    delegate.write(b)
    handleByte(b.toByte())
  }

  override fun write(b: ByteArray) {
    delegate.write(b)
    for (byte in b) handleByte(byte)
  }

  override fun write(b: ByteArray, off: Int, len: Int) {
    delegate.write(b, off, len)
    for (i in off until off + len) handleByte(b[i])
  }

  override fun flush() {
    delegate.flush()
  }

  override fun close() {
    if (lineBuffer.isNotEmpty()) {
      processLine(lineBuffer.toString())
      lineBuffer.clear()
    }
    delegate.close()
  }

  private fun handleByte(b: Byte) {
    val c = (b.toInt() and 0xFF).toChar()
    if (c == '\n' || c == '\r') {
      if (lineBuffer.isNotEmpty()) {
        processLine(lineBuffer.toString())
        lineBuffer.clear()
      }
    } else {
      lineBuffer.append(c)
      // Bound the buffer so a pathological no-newline stream can't OOM us. The
      // marker line is < 1 KiB in practice; 16 KiB leaves ample headroom and
      // matches a typical compiler diagnostic length.
      if (lineBuffer.length > MAX_LINE_LENGTH) {
        processLine(lineBuffer.toString())
        lineBuffer.clear()
      }
    }
  }

  private fun processLine(line: String) {
    val match = STORAGE_REGEX.find(line) ?: return
    val dir = findCachesJvmAncestor(File(match.groupValues[1])) ?: return
    detected.add(dir)
  }

  companion object {
    private const val MAX_LINE_LENGTH = 16 * 1024
    private val STORAGE_REGEX = Regex("""Storage for \[([^\]]+)] is already registered""")

    /**
     * Walks parents of [path] until a directory named `caches-jvm` is found. Returns null if no
     * such ancestor exists — defensive against future Kotlin error-message refactors that change
     * the cache layout.
     */
    fun findCachesJvmAncestor(path: File): File? {
      var current: File? = path
      while (current != null) {
        if (current.name == "caches-jvm") return current
        current = current.parentFile
      }
      return null
    }
  }
}
